presto-main/src/main/java/com/facebook/presto/operator/ParallelHashBuilder.java

```diff
@@ -270,7 +271,10 @@ public void addInput(Page page)
                 size += index.getEstimatedSize().toBytes();
             }
 
-            operatorContext.setMemoryReservation(size);
+            // PagesIndex.addPage doesn't track page memory but blocks are shared and kept
+            pagesMemorySize += page.getRetainedSizeInBytes();
```

Comment:
Shouldn't this be accounted in `PagesIndex.getEstimatedSize()` ? 

Ideally this same memory reservation from those PageIndexes should be later "transferred" to InMemoryJoinHash. Otherwise I would be afraid about various race conditions between finishing `ParallelHashCollectOperator` and starting `ParallelHashBuilderOperator` - memory allocation for pages could end up being counted zero, once or twice. I am not familiar with Presto's memory tracking mechanisms, so I don't know if that's indeed problematic scenario and if it's easy to fix.